### PR TITLE
Feature/apollo federation 2

### PR DESCRIPTION
**Pull request by mtskillman**
Add support for the new Apollo Federation 2 directives, with the exception of the "@link" directive. Also modify "@key" directive to account for Apollo Federation 2 changes. 

## Description
    1. adjust imports (such as the "from .schema_directives import Key" in strawberry/federation/schema.py) to account for circular dependency issue
    2. add typing for the shareable, tag, override, inaccessible directives
    3. remove "link" directive (will be addressed by some other PR)
    4. add various test cases for the new directives
    5. update "key" directive test case to account for the "resolvable" property
    6. new schema directives (@shareable, @tag, @override, and @inaccessible) added
    7. adjust @key directive to align 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/strawberry-graphql/strawberry/issues/1827

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
# Pre-release
:wave:

Releasing commit [e63f7a273a02bc86fa6e583d413494d4240c809a] to PyPi as pre-release! :package:

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.111.0.dev.1651525459** [e63f7a273a02bc86fa6e583d413494d4240c809a] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.111.0.dev.1651525459
```

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release adds support for Apollo Federation 2 directives:
- @shareable
- @tag
- @override
- @inaccessible

This release does **not** add support for the @link directive.

This release updates the @key directive to align with Apollo Federation 2 updates.

See the below code snippet and/or the newly-added test cases for examples on how to use the new directives.
The below snippet demonstrates the @override directive.
```python
import strawberry
from typing import List

@strawberry.interface
class SomeInterface:
    id: strawberry.ID

@strawberry.federation.type(keys=["upc"], extend=True)
class Product(SomeInterface):
    upc: str = strawberry.federation.field(external=True, override=["mySubGraph"])

@strawberry.federation.type
class Query:
    @strawberry.field
    def top_products(self, first: int) -> List[Product]:
        return []

schema = strawberry.federation.Schema(query=Query)
```

should return:

```graphql
extend type Product implements SomeInterface @key(fields: "upc", resolvable: "True") {
  id: ID!
  upc: String! @external @override(from: "mySubGraph")
}

type Query {
  _service: _Service!
  _entities(representations: [_Any!]!): [_Entity]!
  topProducts(first: Int!): [Product!]!
}

interface SomeInterface {
  id: ID!
}

scalar _Any

union _Entity = Product

type _Service {
  sdl: String!
}
```



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Matt Skillman for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/federation/field.py [hunk 1]
```diff
@@ -52,6 +50,10 @@ def field(
     provides: Optional[List[str]] = None,
     requires: Optional[List[str]] = None,
     external: bool = False,
+    shareable: bool = False,
+    tag: Optional[List[str]] = None,
+    override: Optional[List[str]] = None,
```
Comment (patrick91) [inline strawberry/federation/field.py:55]:
```suggestion
    tags: Optional[Iterable[str]] = (),
    overrides: Optional[Iterable[str]] = (),
```

File: strawberry/federation/schema_directives.py [hunk 2]
```diff
@@ -23,3 +28,42 @@ class Provides:
...
+
+
+@schema_directive(
+    locations=[Location.FIELD_DEFINITION, Location.OBJECT], name="shareable"
+)
+class Shareable:
+    ...
+
+
+@schema_directive(locations=[Location.FIELD_DEFINITION], name="tag")
```
Comment (patrick91) [inline strawberry/federation/schema_directives.py:49]: this is supported on all these:

```
FIELD_DEFINITION | INTERFACE | OBJECT | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
```

source: https://cs.github.com/apollographql/federation/blob/7331cbdb824af20f70ca500f1826859c92edda49/federation-integration-testsuite-js/src/fixtures/reviews.ts?q=directive+%40tag#L9-L19

Comment (mtskillman) [reply strawberry/federation/schema_directives.py]: resolved

File: strawberry/federation/field.py (hunk 1 above)
Comment (mtskillman) [reply strawberry/federation/field.py]: resolved, with the exception of "overrides" as it is singular rather than plural -- not repeatable (or something)

**Review by patrick91 — APPROVED** [review]

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**